geom/type_sequence.go

```diff
@@ -0,0 +1,145 @@
+}
+
+// CoordinatesType returns the coordinates type used to represent point
+// locations in the Sequence.
+func (s Sequence) CoordinatesType() CoordinatesType {
+	return s.ctype
+}
+
+// Length returns the number of point locations represented by the Sequence.
+func (s Sequence) Length() int {
+	return len(s.floats) / s.ctype.Dimension()
+}
+
+// Get returns the Coordinates of the ith point location in the Sequence. It
+// panics if i is out of range with respect to the number of points in the
+// Sequence.
+func (s Sequence) Get(i int) Coordinates {
+	stride := s.ctype.Dimension()
+	c := Coordinates{
+		XY: XY{
+			s.floats[i*stride],
+			s.floats[i*stride+1],
+		},
+		Type: s.ctype,
+	}
+	switch s.ctype {
+	case DimXYZ:
+		c.Z = s.floats[i*stride+2]
+	case DimXYM:
+		c.M = s.floats[i*stride+2]
+	case DimXYZM:
+		c.Z = s.floats[i*stride+2]
+		c.M = s.floats[i*stride+3]
+	}
+	return c
+}
+
+// GetXY returns the XY of the ith point location in the Sequence. It panics if
+// i is out of range with respect to the number of points in the Sequence.
+func (s Sequence) GetXY(i int) XY {
+	stride := s.ctype.Dimension()
+	return XY{
+		s.floats[i*stride],
+		s.floats[i*stride+1],
+	}
+}
+
+// Reverse returns a new Sequence containing the same point locations, but in
+// reversed order.
+func (s Sequence) Reverse() Sequence {
+	stride := s.ctype.Dimension()
+	n := s.Length()
+	reversed := make([]float64, len(s.floats))
+	for i := 0; i < n; i++ {
+		j := n - i - 1
+		copy(
+			reversed[i*stride:(i+1)*stride],
+			s.floats[j*stride:(j+1)*stride],
+		)
+	}
+	return Sequence{s.ctype, reversed}
+}
+
+// ForceCoordinatesType returns a new Sequence with a different CoordinatesType. If a
+// dimension is added, then its new value is set to zero for each point
+// location in the Sequence.
+func (s Sequence) ForceCoordinatesType(newCType CoordinatesType) Sequence {
+	if s.ctype == newCType {
+		return s
+	}
+	if len(s.floats) == 0 {
+		return Sequence{newCType, nil}
+	}
+
+	n := s.Length()
+	stride := newCType.Dimension()
+	flat := make([]float64, stride*n)
+	for i := 0; i < n; i++ {
+		c := s.Get(i)
+		flat[stride*i+0] = c.X
+		flat[stride*i+1] = c.Y
+		switch newCType {
+		case DimXYZ:
+			flat[stride*i+2] = c.Z
+		case DimXYM:
+			flat[stride*i+2] = c.M
+		case DimXYZM:
+			flat[stride*i+2] = c.Z
+			flat[stride*i+3] = c.M
+		}
+	}
+	return Sequence{newCType, flat}
+}
+
+// Force2D returns a new Sequence with Z and M values removed (if present).
+func (s Sequence) Force2D() Sequence {
+	return s.ForceCoordinatesType(DimXY)
+}
+
+// getLine extracts a 2D line segment from a sequence by joining together
+// adjacent locations in the sequence. It is designed to be called with i equal
+// to each index in the sequence (from 0 to n-1). The flag indicates if the
+// returned line is valid.
+func getLine(seq Sequence, i int) (Line, bool) {
+	if i == 0 {
+		return Line{}, false
+	}
+	ln := Line{
+		a: Coordinates{Type: DimXY, XY: seq.GetXY(i - 1)},
```

Comment:
Does this usage match the function comment?

Reply:
I think it does. The idea is that the caller should call the function with `i` equal to each index of the sequence. The trouble is that there are `n-1` lines (or less in the case of duplicated points) in a sequence. So some of the values of `i` will have to return `false`. I wanted to leave that as an implementation detail of this function that the caller shouldn't have to worry about. That way we're free to change the implementation if there's a good reason to without having to change the callers (which we might have to do if the contract of the function was that the caller shouldn't call the function with `i=0`).

Reply:
OK, I just wanted to ensure you were being explicitly vague about whether `i==0` or `i==1` yields the first line. And as you say, duplicates could make the mapping even less clear. I'm just too used to seeing zero meaning the first element, and of course in a C/C++ world where the number of elements in an array might be construed as a `size_t` or some other `unsigned` type, the use of `i-1` looked suspicious (but is actually perfectly fine here).